### PR TITLE
[AERIE-1847]  Add duration filter to constraints eDSL

### PR DESCRIPTION
**Pull request by adrienmaillard**
* **Tickets addressed:** AERIE-1847
* **Review:** By commit  
* **Merge strategy:** Merge (no squash)  

## Description

This is building on top of #179.

I have reimplemented the filters in the constraints library. 

## Verification

Several tests have been written/modified. 

## Documentation

Documentation about these two constraints has been added to the [Constraints](https://github.com/NASA-AMMOS/aerie/wiki/Constraints) page.

## Future work

All of the other filters/transformers.

### CODE REVIEW

File: merlin-server/constraints-dsl-compiler/src/libs/constraints-edsl-fluent-api.ts [hunk 1]
```diff
@@ -83,6 +83,22 @@ export class Windows {
       expression: this.__astNode,
     });
   }
+
+  public longerThan(duration: number) : Windows {
```
Comment (mattdailis) [inline merlin-server/constraints-dsl-compiler/src/libs/constraints-edsl-fluent-api.ts]: Let's use the Duration type for durations (even though it is currently aliased to `number`)

```suggestion
  public longerThan(duration: Duration) : Windows {
```

File: merlin-server/constraints-dsl-compiler/src/libs/constraints-edsl-fluent-api.ts [hunk 2]
```diff
@@ -83,6 +83,22 @@ export class Windows {
...
+    return new Windows({
+      kind: AST.NodeKind.WindowsExpressionLongerThan,
+      windowExpression: this.__astNode,
+      duration: duration
+    })
+  }
+
+  public shorterThan(duration: number) : Windows {
```
Comment (mattdailis) [inline merlin-server/constraints-dsl-compiler/src/libs/constraints-edsl-fluent-api.ts]:
```suggestion
  public shorterThan(duration: Duration) : Windows {
```

File: merlin-server/constraints-dsl-compiler/src/libs/constraints-edsl-fluent-api.ts [hunk 3]
```diff
@@ -327,6 +343,10 @@ declare global {
      * this method to produce a violation whenever it is NOT satisfied.
      */
     public violations(): Constraint;
+
+    public longerThan(duration: number): Windows;
+    public shorterThan(duration: number): Windows;
```
Comment (mattdailis) [inline merlin-server/constraints-dsl-compiler/src/libs/constraints-edsl-fluent-api.ts]:
```suggestion
    public longerThan(duration: Duration): Windows;
    public shorterThan(duration: Duration): Windows;
```

**Review by mattdailis — APPROVED** [review]
Looks like there are some extra files (`README` and `windows-edsl-fluent-api.ts`), but I _think_ they'll disappear after a rebase. If not, we should delete them.

Looks good to me! I'm wondering if in the DSL fluent API, if there's any clearer word than "longerThan". I'm thinking along the lines of `forADurationOfAtLeast` and `forADurationOfAtMost` (too long, I know 😝 )

Here's how it would look with `forLongerThan`:
```typescript
Real.Resource("state of charge").rate().equal(Real.Value(4.0)).forLongerThan(1000)
```
it reads well in that example, but I'm not sure if it will make grammatical sense in all cases 🤷 . Just brainstorming, I'm happy to leave the names as is and rename them if ever we decide we've found better ones.

**Comment by adrienmaillard** [comment]
Weird stuff: the branch has been rebased onto develop, then forced-pushed. On my side, the branch is clean, only containing my 3 commits. Still got a small part of 4 of commits from your branch.

**Comment by adrienmaillard** [comment]
Thank you for the approval. I made the changes you asked. If we change the names, I'd be in favor of your longer option with `duration` inside the name to make clear we go from the resource domain to the window domain.

**Review by goetzrrGit — APPROVED** [review]

**Review by patkenneally — CHANGES_REQUESTED** [review]
Has the documentation been updated? If/when it has please update the PR description and we can merge.

**Comment by adrienmaillard** [comment]
@patkenneally I've added some documentation on the https://github.com/NASA-AMMOS/aerie/wiki/Constraints page.

**Review by patkenneally — APPROVED** [review]

**Pull request merged**